tokio/src/sync/semaphore.rs

```diff
@@ -990,6 +990,24 @@ impl<'a> SemaphorePermit<'a> {
         self.permits += other.permits;
         other.permits = 0;
     }
+
+    /// Detaches `n` permits from `self` and returns a new [`SemaphorePermit`] instance that holds `n` permits.
+    ///
+    /// It guarantees at least one permit held by both `self` and the new instance.
```

Comment:
I feel that allowing `detach(0)` (or `detach(n)` for a `SemaphorePermit` with `n` permits) could simplify the API. In fact, it's technically possible to create a zero-permit `SemaphorePermit` with `try_acquire_many(0)` even now.
Have you experienced scenarios where you want the permit to always be more than 1?

Reply:
I limited this because the zero-permit `SemaphorePermit` usually makes no sense, and we have no way to tell the user how many permits are held by a `SemaphorePermit`.

What about adding a `held_permits` (or a better name) method to return the number of permits held by a `SemaphorePermit` and removing these restrictions?

Reply:
If all of the `*acquire_many` methods already allow zero permits, then I see no harm in allowing it here. I would perhaps all the method `num_permits`?